omx/gstomxvideofilter.c

```diff
@@ -0,0 +1,2398 @@
+  PROP_OUTPUT_BUFFERS,
+  PROP_INPUT_BUFFERS
+};
+
+#define GST_OMX_VIDEO_FILTER_ALWAYS_COPY_DEFAULT FALSE
+#define GST_OMX_VIDEO_FILTER_OUTPUT_BUFFERS_DEFAULT 6
+#define GST_OMX_VIDEO_FILTER_INPUT_BUFFERS_DEFAULT 6
+
+/* we can't use G_DEFINE_ABSTRACT_TYPE because we need the klass in the _init
+ * method to get to the padtemplates */
+GType
+gst_omx_video_filter_get_type (void)
+{
+  static volatile gsize type = 0;
+
+  if (g_once_init_enter (&type)) {
+    GType _type;
+    static const GTypeInfo info = {
+      sizeof (GstOMXVideoFilterClass),
+      NULL,
+      NULL,
+      (GClassInitFunc) gst_omx_video_filter_class_init,
+      NULL,
+      NULL,
+      sizeof (GstOMXVideoFilter),
+      0,
+      (GInstanceInitFunc) gst_omx_video_filter_init,
+    };
+    const GInterfaceInfo preset_interface_info = {
+      NULL,                     /* interface_init */
+      NULL,                     /* interface_finalize */
+      NULL                      /* interface_data */
+    };
+
+    _type = g_type_register_static (GST_TYPE_ELEMENT,
+        "GstOMXVideoFilter", &info, G_TYPE_FLAG_ABSTRACT);
+    g_type_add_interface_static (_type, GST_TYPE_PRESET,
+        &preset_interface_info);
+    g_once_init_leave (&type, _type);
+  }
+  return type;
+}
+
+
+static void
+gst_omx_video_filter_class_init (GstOMXVideoFilterClass * klass)
+{
+  GObjectClass *gobject_class = G_OBJECT_CLASS (klass);
+  GstElementClass *element_class = GST_ELEMENT_CLASS (klass);
+
+  GST_DEBUG_CATEGORY_INIT (gst_omx_video_filter_debug_category,
+      "omxvideofilter", 0, "OMX video filter base class");
+
+  parent_class = g_type_class_peek_parent (klass);
+
+  g_type_class_add_private (klass, sizeof (GstOMXVideoFilterPrivate));
+
+  gobject_class->finalize = gst_omx_video_filter_finalize;
+  gobject_class->set_property = gst_omx_video_filter_set_property;
+  gobject_class->get_property = gst_omx_video_filter_get_property;
+
+  g_object_class_install_property (gobject_class, PROP_ALWAYS_COPY,
+      g_param_spec_boolean ("always-copy", "Always copy",
+          "If the buffer will be used or not directly for the OpenMax component",
+          GST_OMX_VIDEO_FILTER_ALWAYS_COPY_DEFAULT, G_PARAM_WRITABLE));
+
+  g_object_class_install_property (gobject_class, PROP_OUTPUT_BUFFERS,
+      g_param_spec_uint ("output-buffers", "Output buffers",
+          "The amount of OMX output buffers",
+          1, 16, GST_OMX_VIDEO_FILTER_OUTPUT_BUFFERS_DEFAULT,
+          G_PARAM_READWRITE));
+
+  g_object_class_install_property (gobject_class, PROP_INPUT_BUFFERS,
+      g_param_spec_uint ("input-buffers", "Input buffers",
+          "The amount of OMX input buffers",
+          1, 16, GST_OMX_VIDEO_FILTER_INPUT_BUFFERS_DEFAULT,
+          G_PARAM_READWRITE));
+
+  element_class->change_state =
+      GST_DEBUG_FUNCPTR (gst_omx_video_filter_change_state);
+
+
+  klass->cdata.type = GST_OMX_COMPONENT_TYPE_FILTER;
+  klass->cdata.default_sink_template_caps = "video/x-raw, "
```

Comment:
Install new pad templates

Reply:
The code in gstomx.c requires the template caps defined by this way, since they complement the information extracted from the config file (gstomx.conf), if these templates are not in the cdata that logic fails and would need refactoring.